### PR TITLE
Fix Embedding.get_delta_weight replacing nn.Parameter with plain tensor

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
## Summary

- `Embedding.get_delta_weight` replaces `nn.Parameter` objects in `lora_embedding_A`/`lora_embedding_B` with plain tensors when running on CPU with float16/bfloat16 weights
- This breaks gradient tracking, parameter registration, and optimizer state for any training that follows a merge/unmerge cycle
- Fix uses `.data` assignment to preserve the `nn.Parameter` wrapper, consistent with `Linear.get_delta_weight` and `_ConvNd.get_delta_weight` which already handle this correctly

## Details

When `cast_to_fp32` is `True`, the current code does:

```python
weight_A = self.lora_embedding_A[adapter]  # nn.Parameter
weight_A = weight_A.float()                # plain tensor (detached)
# ... compute output ...
self.lora_embedding_A[adapter] = weight_A.to(dtype)  # replaces Parameter with tensor!
```

The `.float()` call creates a new plain tensor. Assigning it back directly overwrites the `nn.Parameter`. The fix:

```python
self.lora_embedding_A[adapter].data = weight_A.to(dtype)
self.lora_embedding_B[adapter].data = weight_B.to(dtype)
```

This matches the pattern used in `Linear.get_delta_weight` (line 893-894) and `_ConvNd.get_delta_weight` (line 1498-1499).

## Test plan

- [ ] Verify `Embedding.get_delta_weight` preserves `nn.Parameter` type after call on CPU with float16 weights
- [ ] Verify merge/unmerge cycle on LoRA-adapted embedding layers does not break subsequent training
- [ ] Existing tests continue to pass

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks, this is a nice catch. Could you please also provide a unit test for this bug? There isn't really a perfect place to put it, I'd suggest to add a one-off test to `test_decoder_models.py` using one of the toy decoder models and targeting the embedding layer to reproduce the behavior. LMK if you need further guidance.

**Comment by BenjaminBossan** [comment]
Thanks for the unit test, it looks quite good. I ran it locally on the main branch and to my surprise, it passed even without your fix. Jumping into the debugger, I see:

```python
1136 	            self.lora_embedding_A[adapter] = weight_A.to(dtype)
1137 ->	            breakpoint()
1138 	            self.lora_embedding_B[adapter] = weight_B.to(dtype)
1139 	
1140 	        return output_tensor
(Pdb) type(self.lora_embedding_B[adapter])
<class 'torch.nn.parameter.Parameter'>
(Pdb) type(weight_B.to(dtype))
<class 'torch.Tensor'>
(Pdb) self.lora_embedding_B[adapter] = weight_B.to(dtype)
(Pdb) type(self.lora_embedding_B[adapter])
<class 'torch.nn.parameter.Parameter'>
```

So it appears like `nn.ParameterDict` automatically converts the tensor to an `nn.Parameter` on assignment, which I didn't know. Did you observe any faulty behavior in the wild?

**Comment by Mr-Neutr0n** [comment]
you're right — nn.ParameterDict handles the conversion automatically. didn't realize that. closing since the fix isn't actually needed. thanks for digging into it\!
